### PR TITLE
Problem: lack conversion `message_t` and `multipart_t`

### PR DESCRIPTION
**Pull request by brettviren**
Solution: add encode/decode methods to `multipart_t` giving a codec compatible with the CZMQ equivalent.

This is part of what was initially #388.

### CODE REVIEW

**Comment by gummif** [comment]
Have you considered implementing these as free functions similar to send_multipart? That would allow constructing the message without creating a multipart_t object.

Also, is the protocol documented somewhere?

**Comment by brettviren** [comment]
Hi @gummif.  I did consider it but seeing all the other helper methods in `multipart_t` that add value to `m_parts` I figured having `encode()/decode()` as methods fit into the current convention.

I agree what you say about a free function would not require a `multipart_t` to be created if the user doesn't need one.  OTOH, a free function means users that **do** want a `multipart_t` need an extra step.

What do you think about adding free functions in addition to the methods?  These functions could hold the actual codec code and be called by the methods.  Furthermore the functions could be templated to accept an iterator range so not tied to a particular container.

I haven't found anywhere in ZeroMQ ecosystem that documents the codec.  It's very simple and I figured it out by reading the few lines of `zmsg_encode()/_decode()` in CZMQ.   I do happen to have some blurb in the README of the package that currently drives the need for this codec and I reproduce it below:

> The encoding is compatible with the CZMQ `zmsg_encode()` function. Specifically, the data of small messages less than 255 bytes are prefixed with a 1-byte size. Larger messages are prefixed by a single byte `0xFF` followed by a 4-byte size value.

BTW, I have a Python implementation for this as well:

https://github.com/brettviren/generaldomo/blob/master/generaldomo/zhelpers.py#L75

**Comment by gummif** [comment]
Free functions + member functions for convenience sounds good.

I wonder if this should be officially documented (in libzmq maybe), especially since the new SERVER/CLIENT socket types can only handle single part messages.

**Comment by brettviren** [comment]
> Free functions + member functions for convenience sounds good

So, how would this additional feature be handled?  Does this PR go through and adding the free functions would be another PR (by you maybe?) or does this PR gain a second commit (by me) or do I need to withdraw this PR and make yet another?

**Comment by gummif** [comment]
I guess you can choose. Either continue and  implement it in this PR, with new commits or rebasing, or that is a feature that can be added later in a second PR.

File: zmq_addon.hpp [hunk 1]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+    bool decode(const message_t& encoded) {
+        const unsigned char *source = encoded.data<unsigned char>();
+        const unsigned char *limit = source + encoded.size();
+        multipart_t decoded;
+        while (source < limit) {
+            size_t part_size = *source++;
+            if (part_size == 255) {
+                if (source > limit - 4) {
+                    // garbage, bail
+                    return false;
```
Comment (gummif) [inline zmq_addon.hpp]: I wonder if throwing an exception (`std::argument_error` maybe) would be better, since that should be rare and not ignored.

File: tests/multipart.cpp [hunk 2]
```diff
@@ -178,4 +178,39 @@ TEST_CASE("multipart legacy test", "[multipart]")
...
 }
+
+TEST_CASE("multipart codec", "[multipart]")
+{
+    using namespace zmq;
+
+    multipart_t mmsg;
+    mmsg.addstr("Hello World");
+    message_t msg = mmsg.encode();
+    assert(msg.size() == 1 + 11); // small size packing
```
Comment (gummif) [inline tests/multipart.cpp]: The CHECK or REQUIRE macro should be used for unit tests.

Looks like the older unit tests in the file need a bit a refactoring.

**Comment by sigiesec** [comment]
> Free functions + member functions for convenience sounds good.

Agree :)

> I wonder if this should be officially documented (in libzmq maybe), especially since the new SERVER/CLIENT socket types can only handle single part messages.

I think the appropriate place to specify this independently from the czmq implementation would be a ZeroMQ RFC (https://rfc.zeromq.org/). libzmq only provides the low-level API, which does not deal with this (which is why I was concerned about adding this to cppzmq, which wraps the low-level libzmq API). @bluca WDYT? Does it make sense to add a RFC for the multipart encoding?

File: zmq_addon.hpp [hunk 3]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+        for (auto& part : *this) {
+            size_t part_size = part.size();
+            size_t count_size = 5;
+            if (part_size < 255) {
+                count_size = 1;
+            }
+            mmsg_size += part_size + count_size;
+        }
+
+        message_t encoded(mmsg_size);
```
[the remark this answers is not in the document]
Comment (sigiesec) [reply zmq_addon.hpp]: Hm, I don't think so. It's a corner case, but why should it be invalid to have a 0-part multi-part message?

But I don't know if `zmsg_encode` allows that. It should be done like that function does, since the goal is to be compatible.

@brettviren Please double-check and confirm this. In any case, please add a test case for this corner case.

File: zmq_addon.hpp [hunk 4]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+        unsigned char* buf = encoded.data<unsigned char>();
+        for (auto& part : *this) {
+            size_t part_size = part.size();
```
[the remark this answers is not in the document]
Comment (sigiesec) [reply zmq_addon.hpp]: Yes, that should be checked. I think a run-time exception would be more appropriate here.

File: zmq_addon.hpp (hunk 1 above)
Comment (sigiesec) [reply zmq_addon.hpp]: I also think that an exception is preferrable for this case, in line with the rest of the API. If there are really use cases, where this might be a more common case, another `try_decode` function might be added that returns a bool in that case. In that case, the return value should be declared nodiscard, when available.

File: zmq_addon.hpp [hunk 5]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+            *buf++ = (part_size >> 24) & 255;
+            *buf++ = (part_size >> 16) & 255;
+            *buf++ = (part_size >>  8) & 255;
+            *buf++ =  part_size        & 255;
+            memcpy (buf, part_data, part_size);
+            buf += part_size;
+        }
+        return encoded;
+    }
+    // Decode the CZMQ zmsg_t/zframe_t compatible encoded single part
```
Comment (sigiesec) [inline zmq_addon.hpp]: nit: please add a blank line before this.

**Comment by bluca** [comment]
Perhaps - CZMQ until now has maintained a few protocols on its own, intended to be used only internally and thus not published. If someone wants to codify those, that should be fine - they change rarely, if ever.

File: zmq_addon.hpp [hunk 6]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+    // message into multi-parts and append to self.  Return true if
```
Comment (sigiesec) [inline zmq_addon.hpp]: The appending behaviour is a bit disputable, and makes me question if having this as a non-static member function is a good idea. Isn't this more error-prone than useful? What about making this a static member function that returns a `multipart_t`?

If this stays like this, there should be at least one test that checks if it appends as expected.

File: zmq_addon.hpp [hunk 7]
```diff
@@ -469,6 +469,79 @@ class multipart_t
...
+    // self was updated.  A false value will be returned if the
+    // encoded is garbage.
+    bool decode(const message_t& encoded) {
+        const unsigned char *source = encoded.data<unsigned char>();
+        const unsigned char *limit = source + encoded.size();
```
Comment (sigiesec) [inline zmq_addon.hpp]: nit: could be declared top-level const
```suggestion
        const unsigned char * const limit = source + encoded.size();
```

File: tests/multipart.cpp [hunk 8]
```diff
@@ -178,4 +178,39 @@ TEST_CASE("multipart legacy test", "[multipart]")
     assert(received.empty());
     assert(str == "One-hundred");
 }
+
+TEST_CASE("multipart codec", "[multipart]")
```
Comment (sigiesec) [inline tests/multipart.cpp]: This seems to contain two test cases that somehow depend on each other. Please split this up, and consider this when adding further test cases.

File: zmq_addon.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (brettviren) [reply zmq_addon.hpp]: So far I have yet to require sending an empty multipart-encoded in any higher-level protocols I've used.  Certainly, if it can be done it will be done eventually.

I think the proper thing is for an empty `multipart_t` to encode to the default (empty) `message_t` and vice versa. 

I'll double check what CZMQ does here and add some tests for this case.

File: zmq_addon.hpp (hunk 1 above)
Comment (brettviren) [reply zmq_addon.hpp]: Sounds good.  Consider this and the above added to my list.

File: tests/multipart.cpp (hunk 2 above)
Comment (brettviren) [reply tests/multipart.cpp]: I'll leave the older test to others but I will re-instrument this test with macros as suggested.

File: zmq_addon.hpp (hunk 6 above)
Comment (brettviren) [reply zmq_addon.hpp]: I don't fully understand this comment.  Are you worried the user may allow a `decode()` to append to a non-empty `multipart_t`?  If so I vacillated on the choice to allow it.  But, the opposite idiom is used in `multipart_t::recv()` so maybe I chose badly.  

If the comment is about something else, please elaborate.

File: tests/multipart.cpp (hunk 8 above)
Comment (brettviren) [reply tests/multipart.cpp]: Will do.  They are actually independent.  I'll add the others tests suggested as individual tests.

**Comment by brettviren** [comment]
Thanks for all the comments.  One clarification requested and the rest all seem good to me.

File: zmq_addon.hpp (hunk 6 above)
Comment (gummif) [reply zmq_addon.hpp]: Yes, should it append or replace the content. Replacing is probably more common or expected behavior. Or as suggested use  `multipart_t mm = multipart_t::decode(my_message);`.

**Comment by brettviren** [comment]
@gummif I've been able to make `encode()` a free function templated on a Range based on your `send_multipart()`.  To do similar for `decode()` I need a way to fill a `multipart_t` in a `std::` container looking manner.  I see I can mimic what you did with `recv_multipart()` it seems this function can not be used to directly fill a `multipart_t`. 

To make `recv_multipart()` and a free `decode()` work, I think `multipart_t` needs to be changed to look more like an STL container.  I'm almost at the edge of my C++ abilities, but I'll try to minimally add to `multipart_t` what's needed.  If you or others have any immediate suggestions, please let me know. 

I will appreciate another round of careful reviewing after I update the PR.

**Comment by brettviren** [comment]
Turns out only minor additions to `multipart_t` are needed to let it be used with a `std::back_inserter` to provide `std::copy()` like calling for `decode()`.  I didn't try but I think this gives the knock-on benefit of letting `recv_multipart()` to actually be used with a `multipart_` (unless I miss this already working).

I removed the Boolean returns in favor of throwing.  I picked `std::range_error` if absurdly large parts are used (more than 32 bits) and `std::out_of_range` if encoded data is garbage and would otherwise lead to overflow errors.

I double checked what CZMQ does with empty multiparts and indeed it does what we do here.

I moved the codec tests to their own file and added more to cover the free functions empties, tickling the out of range exception.  I also made and ran a test that exercises the range error with a 4GB+ part but I disabled it as I worry about false negative failures when run on low-memory hosts (travis-ci?).

I also fixed the nits and ran clang-format.

And as I write this I see travis-ci is failing.... Of course it all is clean on my system.  Looks like a compilation error on one version of GCC and not related directly to the changes.  I'll fix this and see if I can squash that fix to what I just pushed to keep each commit in this PR green.

**Comment by brettviren** [comment]
Yay, green.  Please have a look at 3191ae6 @sigiesec and @gummif and anyone else interested.

File: tests/multipart.cpp [hunk 9]
```diff
@@ -178,4 +178,5 @@ TEST_CASE("multipart legacy test", "[multipart]")
     assert(received.empty());
     assert(str == "One-hundred");
 }
+
```
Comment (sigiesec) [inline tests/multipart.cpp]: Stray change here, please remove this such that tests/multipart.cpp is not touched at all.

File: tests/codec_multipart.cpp [hunk 10]
```diff
@@ -0,0 +1,168 @@
+#include <catch.hpp>
+#include <zmq_addon.hpp>
+
+#ifdef ZMQ_HAS_RVALUE_REFS
```
Comment (sigiesec) [inline tests/codec_multipart.cpp]: Do all of these tests require rvalue references?

File: zmq_addon.hpp [hunk 11]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+template<class Range
+#ifndef ZMQ_CPP11_PARTIAL
+         ,
+         typename = typename std::enable_if<
+           detail::is_range<Range>::value
+           && (std::is_same<detail::range_value_t<Range>, message_t>::value
+               || detail::is_buffer<detail::range_value_t<Range>>::value)>::type
 #endif
+         >
+message_t encode(const Range &parts)
```
Comment (gummif) [inline zmq_addon.hpp]: Do you think `encode_multipart` or `multipart_encode` (and `decode_multipart` or `multipart_decode`) would be more descriptive?

File: zmq_addon.hpp [hunk 12]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+{
+    size_t mmsg_size = 0;
+
+    // First pass check sizes
+    for (const message_t &part : parts) {
```
Comment (gummif) [inline zmq_addon.hpp]: To allow `const_buffer` and `mutable_buffer` to be used also, you can change this to `const auto&`, and use `static_cast<const unsigned char*>(part.data())` instead of `part.data<const unsigned char>()` below.

File: zmq_addon.hpp [hunk 13]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+        size_t part_size = part.size();
+        if (part_size > 0xFFFFFFFF) {
```
Comment (gummif) [inline zmq_addon.hpp]: Consider using `std::numeric_limits<uint32_t>::max()`

File: zmq_addon.hpp [hunk 14]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+            // Size value must fit into uint32_t.
+            throw std::range_error("Invalid size, message part too large");
+        }
+        size_t count_size = 5;
+        if (part_size < 255) {
```
Comment (gummif) [inline zmq_addon.hpp]: Consider using `std::numeric_limits<uint8_t>::max()`

File: zmq_addon.hpp [hunk 15]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+
+    while (source < limit) {
+        size_t part_size = *source++;
+        if (part_size == 255) {
+            if (source > limit - 4) {
+                throw std::out_of_range(
+                  "Malformed encoding, overflow in reading size");
+            }
+            part_size =
+              (source[0] << 24) + (source[1] << 16) + (source[2] << 8) + source[3];
```
Comment (gummif) [inline zmq_addon.hpp]: The types here are `unsigned char` and `int`, so I think there could be overflow if `source[0]` is big enough. There should probably by casts to `uint32_t`.

File: zmq_addon.hpp [hunk 16]
```diff
@@ -154,7 +155,122 @@ send_multipart(socket_ref s, Range &&msgs, send_flags flags = send_flags::none)
...
+            source += 4;
+        }
+
+        if (source > limit - part_size) {
+            throw std::out_of_range("Malformed encoding, overflow in reading part");
+        }
+        message_t msg(source, part_size);
+        *out = std::move(msg);
```
Comment (gummif) [inline zmq_addon.hpp]: `*out = message_t(source, part_size);` might be more efficient

File: zmq_addon.hpp [hunk 17]
```diff
@@ -469,6 +590,23 @@ class multipart_t
         return true;
     }
 
+    // Return single part message_t encoded from this multipart_t.
+    message_t encode() const { return zmq::encode(*this); }
```
Comment (gummif) [inline zmq_addon.hpp]: These new functions should be guarded by ZMQ_CPP11

File: tests/codec_multipart.cpp (hunk 10 above)
Comment (gummif) [reply tests/codec_multipart.cpp]: This should probably be ZMQ_CPP11.

File: tests/multipart.cpp (hunk 9 above)
Comment (brettviren) [reply tests/multipart.cpp]: Oops!  Will do.

File: tests/codec_multipart.cpp (hunk 10 above)
Comment (brettviren) [reply tests/codec_multipart.cpp]: Thanks, I'll check this.

File: zmq_addon.hpp (hunk 11 above)
Comment (brettviren) [reply zmq_addon.hpp]: I do not have a strong opinion. but I lean to leaving this as-is or maybe with a different `_qualifier`.  

I'm thinking about future types which may overload `encode()` where the label `_multipart` is not meaningful.  Eg, an application defined  `encode(const MyMessage& mm)` would not necessarily have any "multipartish".

If we had a good name for this codec then `encode_<codecname>()` (etc) would maybe be a better way to distinguish this.

File: zmq_addon.hpp (hunk 12 above)
Comment (brettviren) [reply zmq_addon.hpp]: Excellent, thanks.  I'll try this.

File: zmq_addon.hpp (hunk 15 above)
Comment (brettviren) [reply zmq_addon.hpp]: Indeed!  I confirm if `source[0]` is, eg `0xff` this will not give the expected value.

**Comment by brettviren** [comment]
Latest commit should address all the comments from the previous one except one:  @gummif I still leave the free functions as `encode()/decode()` with my same rational I gave above.  If you feel strongly about using the longer names, please say so and I'll amend.  

Any other comments, @sigiesec or others?  I know it's been tedious to make (and address) them, but the iteration has made this small contribution into much higher quality than it started out as.

**Comment by brettviren** [comment]
appveyor CI failed.  I know nothing about Windows development but it seems to not like the newly added `std::numeric_limits<>::max()`.  I realized I had not included `<limits>` explicitly.  It is maybe included implicitly on gcc but not on Windows?  But, I just force-pushed this fix and it did nothing to help.

I don't have any reasonable way to develop under Windows so I'm not sure what I should do.  Baring a better suggestion, I'll revert back to using literal `0xFF`, etc.

**Comment by sigiesec** [comment]
> appveyor CI failed. I know nothing about Windows development but it seems to not like the newly added `std::numeric_limits<>::max()`. I realized I had not included `<limits>` explicitly. It is maybe included implicitly on gcc but not on Windows? But, I just force-pushed this fix and it did nothing to help.

The windows headers define macros named `min` and `max`... which is horrible. See http://www.suodenjoki.dk/us/archive/2010/min-max.htm, e.g. I am not completely sure, if we should define `NOMINMAX` in the CI script, or in `zmq.hpp` (when on Windows). Since it can't work without defining it (and almost all Windows projects sooner or later define it anyway), the latter seems to be a better idea. So 
```
#ifdef _WIN32
#ifndef NOMINMAX
#define NOMINMAX
#endif
#endif
```

I am not sure what exactly pulls `windows.h` in, so defining it at the very top because the first include directive is probably the best idea.

> 
> I don't have any reasonable way to develop under Windows so I'm not sure what I should do. Baring a better suggestion, I'll revert back to using literal `0xFF`, etc.

**Comment by gummif** [comment]
I think the cause is the winsock2.h inlude in zmq.h (which has been removed on master is seems).

**Comment by gummif** [comment]
Otherwise, looks good to me.

File: zmq_addon.hpp (hunk 12 above)
Comment (brettviren) [reply zmq_addon.hpp]: Support and tests for `const_buffer` and `mutable_buffer` is in the latest commit.

File: zmq_addon.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (brettviren) [reply zmq_addon.hpp]: Indeed CZMQ encodes an empty multipart into an empty single part.  Tests for this case are added.

File: zmq_addon.hpp (hunk 13 above)
Comment (brettviren) [reply zmq_addon.hpp]: Looks like @sigiesec's suggestion for the windows problem let appveyor test pass.  So these use of limits in general looks good now.

**Comment by brettviren** [comment]
Okay, looks all good after telling windows not to be silly about `max`.  I don't know how to clear the "changes requested" but I think everything was addressed.  Let me know if otherwise.

**Comment by brettviren** [comment]
@sigiesec @gummif are you okay to merge this PR?
